server/server.go

```diff
@@ -0,0 +1,388 @@
+package server
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net"
+
+	ipfslite "github.com/hsanjuan/ipfs-lite"
+	blocks "github.com/ipfs/go-block-format"
+	"github.com/ipfs/go-cid"
+	format "github.com/ipfs/go-ipld-format"
+	"google.golang.org/grpc"
+	pb "textile.io/grpc-ipfs-lite/ipfs-lite"
+)
+
+type ipfsLiteServer struct {
+	pb.UnimplementedIpfsLiteServer
+
+	peer *ipfslite.Peer
+}
+
+// StartServer starts the gRPC server
+func StartServer(peer *ipfslite.Peer, host string) error {
+	lis, err := net.Listen("tcp", host)
+	if err != nil {
+		return err
+	}
+
+	grpcServer := grpc.NewServer()
+	server := &ipfsLiteServer{
+		peer: peer,
+	}
+	pb.RegisterIpfsLiteServer(grpcServer, server)
+	return grpcServer.Serve(lis)
+}
+
+type addFileResult struct {
+	Node format.Node
+	Err  error
+}
+
+func addFile(ctx context.Context, peer *ipfslite.Peer, addParams *pb.AddParams, reader io.Reader, ch chan addFileResult) {
+	defer close(ch)
+	params := &ipfslite.AddParams{
+		Layout:    addParams.GetLayout(),
+		Chunker:   addParams.GetChunker(),
+		RawLeaves: addParams.GetRawLeaves(),
+		Hidden:    addParams.GetHidden(),
+		Shard:     addParams.GetShared(),
+		NoCopy:    addParams.GetNoCopy(),
+		HashFun:   addParams.GetHashFun(),
+	}
+	node, err := peer.AddFile(ctx, reader, params)
+	if err != nil {
+		ch <- addFileResult{Err: err}
+		return
+	}
+	ch <- addFileResult{Node: node}
+}
+
+func (s *ipfsLiteServer) AddFile(srv pb.IpfsLite_AddFileServer) error {
+	req, err := srv.Recv()
+	if err != nil {
+		return err
+	}
+	var addParams *pb.AddParams
+	switch payload := req.GetPayload().(type) {
+	case *pb.AddFileRequest_AddParams:
+		addParams = payload.AddParams
+	default:
+		return fmt.Errorf("expexted AddParams for AddFileRequest.Payload but got %T", payload)
+	}
+
+	reader, writer := io.Pipe()
+
+	addFileChannel := make(chan addFileResult)
+	go addFile(srv.Context(), s.peer, addParams, reader, addFileChannel)
+
+	for {
+		req, err := srv.Recv()
+		if err == io.EOF {
+			writer.Close()
+			break
+		} else if err != nil {
+			writer.CloseWithError(err)
+			break
+		}
+		switch payload := req.GetPayload().(type) {
+		case *pb.AddFileRequest_Chunk:
+			_, writeErr := writer.Write(payload.Chunk)
+			if writeErr != nil {
+				return writeErr
+			}
+		default:
+			return fmt.Errorf("expected Chunk for AddFileRequest.Payload but got %T", payload)
+		}
+	}
+
+	addFileResult := <-addFileChannel
```

Comment:
Once the client is done sending data, get the result of the `AddFile` goroutine from the channel.